r/R/expression.R

```diff
@@ -76,7 +76,15 @@
 Expression <- R6Class("Expression", inherit = ArrowObject,
   public = list(
     ToString = function() compute___expr__ToString(self),
-    type = function(schema) compute___expr__type(self, schema),
+    schema = NULL,
+    bind = function(schema) self$schema <- schema,
+    type = function() {
+      if (is.null(self$schema)) {
+        stop("Must bind() expression to a schema before returning its type", call. = FALSE)
+      }
+      compute___expr__type(self, self$schema)
+    },
```

Comment:
@nealrichardson I made this change because we need to call `Expression$type()` in contexts where we don't have access to `schema`.

Let me know if you'd prefer to retain `schema` as an _optional_ argument to `$type()` so that we can get the type either by specifying it in `schema` or passing it to `$bind()` before

Reply:
As you can see, calling the `$bind()` R6 method doesn't actually call the `Bind()` C++ function; I attempted that at first but ran into some problems so switched to doing it like this.